File: src/gorgonetics/genome_parser.py
"""
Shared utilities for parsing genome files and generating gene data.

This module contains common functions used by both the models and template generation scripts.
"""

from __future__ import annotations

from pathlib import Path


def parse_genome_file_header(file_path: str | Path) -> dict[str, str]:
    """
    Parse a genome file and extract header information.

    Args:
        file_path: Path to the genome file

    Returns:
        Dictionary with header information (format_version, breeder, name, genome_type)
    """
    file_path = Path(file_path)

    with open(file_path, encoding="utf-8") as f:
        lines = f.readlines()

    header_info = {"format_version": "", "breeder": "", "name": "", "genome_type": ""}

    in_overview = False

    for line in lines:
        line = line.strip()

        if line == "[Overview]":
            in_overview = True
            continue
        elif line.startswith("["):
            in_overview = False
            continue

        if in_overview and "=" in line:
            key, value = line.split("=", 1)
            key = key.strip()
            value = value.strip()

            if key == "Format":
                header_info["format_version"] = value
            elif key == "Character":
                header_info["breeder"] = value
            elif key == "Entity":
                header_info["name"] = value
            elif key == "Genome":
                header_info["genome_type"] = value

    return header_info


def parse_genome_file_genes(file_path: str | Path) -> dict[str, str]:
    """
    Parse a genome file and extract chromosome gene data.

    Args:
        file_path: Path to the genome file

    Returns:
        Dictionary mapping chromosome numbers to gene data strings
    """
    file_path = Path(file_path)

    with open(file_path, encoding="utf-8") as f:
        lines = f.readlines()

    chromosomes: dict[str, str] = {}
    in_genes = False

    for line in lines:
        line = line.strip()

        if line == "[Genes]":
            in_genes = True
            continue
        elif line.startswith("["):
            in_genes = False
            continue

        if in_genes and "=" in line:
            chr_num, gene_data = line.split("=", 1)
            chr_num = chr_num.strip()
            gene_data = gene_data.strip()
            chromosomes[chr_num] = gene_data

    return chromosomes


def generate_block_letters(num_blocks: int) -> list[str]:
    """
    Generate block letters (A, B, C, ..., Z, AA, AB, ...).

    Args:
        num_blocks: Number of blocks needed

    Returns:
        List of block letter strings
    """
    letters: list[str] = []

    # First 26 blocks use single letters A-Z
    for i in range(min(num_blocks, 26)):
        letters.append(chr(ord("A") + i))

    # If more than 26 blocks, use double letters AA, AB, AC, etc.
    remaining = num_blocks - 26
    if remaining > 0:
        for i in range(remaining):
            first_letter = chr(ord("A") + (i // 26))
            second_letter = chr(ord("A") + (i % 26))
            letters.append(first_letter + second_letter)

    return letters


def count_blocks_in_chromosome(gene_data: str) -> int:
    """
    Count the number of gene blocks in a chromosome.
    Each block is a sequence of gene characters separated by spaces.

    Args:
        gene_data: String containing gene data like "RDRD RDRR ?D?? x?xR"

    Returns:
        Number of blocks
    """
    # Split by spaces and filter out empty strings
    blocks = [block for block in gene_data.split() if block and len(block) >= 2]
    return len(blocks)

File: src/lib/services/genomeParser.ts
/**
 * Genome file parser for Gorgonetics.
 */

import type { Gene, GeneType, Genome } from '$lib/types/index.js';

const VALID_GENE_CHARS = new Set(['R', 'D', 'x', '?']);

/**
 * Parse a genome file's [Overview] section and extract header information.
 */
export function parseGenomeFileHeader(content: string): {
  format_version: string;
  breeder: string;
  name: string;
  genome_type: string;
} {
  const lines = content.split(/\r?\n/);
  const header = { format_version: '', breeder: '', name: '', genome_type: '' };
  let inOverview = false;

  for (const rawLine of lines) {
    const line = rawLine.trim();

    if (line === '[Overview]') {
      inOverview = true;
      continue;
    } else if (line.startsWith('[')) {
      inOverview = false;
      continue;
    }

    if (inOverview && line.includes('=')) {
      const eqIdx = line.indexOf('=');
      const key = line.slice(0, eqIdx).trim();
      const value = line.slice(eqIdx + 1).trim();

      if (key === 'Format') header.format_version = value;
      else if (key === 'Character') header.breeder = value;
      else if (key === 'Entity') header.name = value;
      else if (key === 'Genome') header.genome_type = value;
    }
  }

  return header;
}

/**
 * Parse a genome file's [Genes] section and extract chromosome data.
 * Returns a mapping of chromosome numbers to gene data strings.
 */
export function parseGenomeFileGenes(content: string): Record<string, string> {
  const lines = content.split(/\r?\n/);
  const chromosomes: Record<string, string> = {};
  let inGenes = false;

  for (const rawLine of lines) {
    const line = rawLine.trim();

    if (line === '[Genes]') {
      inGenes = true;
      continue;
    } else if (line.startsWith('[')) {
      inGenes = false;
      continue;
    }

    if (inGenes && line.includes('=')) {
      const eqIdx = line.indexOf('=');
      const chrNum = line.slice(0, eqIdx).trim();
      const geneData = line.slice(eqIdx + 1).trim();
      chromosomes[chrNum] = geneData;
    }
  }

  return chromosomes;
}

/**
 * Convert a block index to a letter (0=A, 1=B, ..., 25=Z, 26=AA, 27=AB, ...).
 */
export function blockLetter(index: number): string {
  if (index < 26) return String.fromCharCode(65 + index);
  const first = Math.floor((index - 26) / 26);
  const second = (index - 26) % 26;
  return String.fromCharCode(65 + first) + String.fromCharCode(65 + second);
}

/**
 * Generate block letters (A, B, C, ..., Z, AA, AB, ...).
 */
export function generateBlockLetters(numBlocks: number): string[] {
  return Array.from({ length: numBlocks }, (_, i) => blockLetter(i));
}

/** Order block letters as `blockLetter` produces them: shorter before longer, lex within length. */
export function compareBlockLetters(a: string, b: string): number {
  return a.length !== b.length ? a.length - b.length : a.localeCompare(b);
}

/**
 * Count the number of gene blocks in a chromosome.
 */
export function countBlocksInChromosome(geneData: string): number {
  return geneData.split(/\s+/).filter((block) => block.length >= 2).length;
}

/**
 * Parse genes for a single chromosome from a gene data string.
 */
export function parseChromosomeGenes(chromosome: string, geneData: string): Gene[] {
  const genes: Gene[] = [];
  const blocks = geneData.split(/\s+/).filter(Boolean);
  const blockLetters = generateBlockLetters(blocks.length);

  for (let blockIdx = 0; blockIdx < blocks.length; blockIdx++) {
    if (blockIdx >= blockLetters.length) break;

    const blockLetter = blockLetters[blockIdx];
    const blockData = blocks[blockIdx];

    for (let posIdx = 0; posIdx < blockData.length; posIdx++) {
      const geneChar = blockData[posIdx];
      if (!VALID_GENE_CHARS.has(geneChar)) continue;

      genes.push({
        chromosome,
        block: blockLetter,
        position: posIdx + 1,
        gene_type: geneChar as GeneType,
      });
    }
  }

  return genes;
}

/**
 * Parse a complete genome from file content.
 */
export function parseGenome(content: string): Genome {
  const header = parseGenomeFileHeader(content);
  const chromosomeData = parseGenomeFileGenes(content);

  const genes: Record<string, Gene[]> = {};
  for (const [chrNum, geneData] of Object.entries(chromosomeData)) {
    const chrNumPadded = chrNum.padStart(2, '0');
    genes[chrNumPadded] = parseChromosomeGenes(chrNumPadded, geneData);
  }

  return {
    format_version: header.format_version,
    breeder: header.breeder,
    name: header.name,
    genome_type: header.genome_type,
    genes,
  };
}

/**
 * Validate that file content looks like a valid genome file.
 */
export function isValidGenomeFile(content: string): boolean {
  return content.includes('[Overview]') && content.includes('[Genes]');
}
